### DISCUSSION TITLE
LORA Trigger Word Position

### DISCUSSION THREAD

**Discussion by DrJimmyBrungus** [DominikDoom/a1111-sd-webui-tagcomplete#232] [General]
Currently when using auto complete to select a LORA (assuming you have the required extension and setting enabled), the list of trigger words are added to the very beginning of the prompt. Would it be possible to have them instead added immediately after the LORA? I feel like the workflow is interrupted when you have to go back to the start of the prompt to search for and delete unwanted trigger words

**Comment by DominikDoom** [comment]
Trigger word position is something I definitely want to add in the future, the initial implementation was just a bit rushed as I was short on time. The decision to put them at the front was because text position matters, so usually if at the beginning they will have more influence as long as you're not using special keywords like BREAK or exceeding the 75 token window.

It's not really selective, but you can also undo the trigger word addition in general by pressing CTRL+Z (or another undo shortcut), as long as you haven't typed / edited anything else after inserting.

**Reply by DrJimmyBrungus** [reply]
Appreciate the super quick reply. Makes sense, I guess if you do implement this change one day, a setting to choose to add it to either the the start or end of the prompt would work best.
Keep up the good work.

**Reply by DominikDoom** [reply]
I had some time to add it in right now, so you're in luck. It was easier than I thought, so I guess it could've been a feature from day one 😅 

https://github.com/DominikDoom/a1111-sd-webui-tagcomplete/assets/34448969/70efbbcc-65ce-495e-9e7d-71a065a15bd3

The options are pretty straightforward, the only thing to note is that with multiple Loras, "Before LORA/LyCO" will only take the current one into account. It won't try to find the first Lora in the prompt. So to add multiple in a row I'd recommend moving the cursor before the first Lora when adding the next, which will keep them at the end. But it shouldn't influence the prompt much either way, since the Lora declarations just get removed during parsing.

**Comment by DrJimmyBrungus** [comment]
Nice work, seems to be working as expected. Regarding multiple LORAs and positioning the cursor, I don't understand the benefit of this. Is there a reason I'd want to do this? Using the Before LORA/LYCO setting, the behaviour is exactly as I'd usually write my prompts.

For example, inserting two LORAs into my prompt:

`tag, tag, <lora1> tag, tag, <lora2>`

I end up with:

`tag, tag, trigger1, trigger1, <lora1> tag, tag, trigger2, trigger2 <lora2>`

which looks fine and normal to me. Should I be doing it another way?

**Reply by DominikDoom** [reply]
It's mainly a matter of personal preferences, some people just like to have all their <lora> definitions at the very end of the prompt.
Like I said before, the only difference with regular models is that words at the beginning will usually have a bit more influence. Other than that there's no relevance to the Lora position since the <bracket> definitions get removed anyway after the specified lora have been loaded.
Same for Lycos and hypernetworks. It's only different with textual inversion embeddings, they will stay where they are since they are essentially just a placeholder/shortcut for concepts and simultaneously definition and trigger word.
